# Hospital dataset

Combining geodata with hospital data.

In [72]:
# Import
import pandas as pd

In [73]:
# Load hospital data
hospitals = pd.read_excel('../data/raw/kzp22_data.xlsx', sheet_name='KZ2022_KZP22', engine='openpyxl')
print(hospitals['Inst'].duplicated().value_counts())
hospitals.info()
hospitals.head()

Inst
False    279
Name: count, dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279 entries, 0 to 278
Columns: 145 entries, JAHR to ErlZvOKPStatVB
dtypes: float64(133), int64(1), object(11)
memory usage: 316.2+ KB


JAHR  KT                         Inst                 Adr  \
0  2022  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1  2022  AG       Kantonsspital Baden AG            Im Ergel   
2  2022  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3  2022  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4  2022  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ RForm Akt       SL              WB  ...  \
0        5000 Aarau  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...   
1        5404 Baden  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...   
2        5000 Aarau  K112    R1   A  NF, IPS       Arzt, BGs  ...   
3  4310 Rheinfelden  K121    R1   A       NF  MSt, Arzt, BGs  ...   
4      5630 Muri AG  K121    R2   A  NF, IPS  MSt, Arzt, BGs  ...   

   AnlZvOKPStatB KostAWLFB KostWBLFB  KostForLFB  ErlAmbStatLangB  ErlOKPAmbB  \
0            NaN       NaN       NaN         NaN              NaN         NaN   
1            NaN       NaN       NaN         NaN              NaN         NaN   
2            NaN       NaN       NaN         NaN              NaN         NaN   
3            NaN       NaN       NaN         NaN              NaN         NaN   
4            NaN       NaN       NaN         NaN              NaN         NaN   

   ErlKVGStatB  ErlKVGStatVB  ErlZvOKPStatB  ErlZvOKPStatVB  
0          NaN           NaN            NaN             NaN  
1          NaN           NaN            NaN             NaN  
2          NaN           NaN            NaN             NaN  
3          NaN           NaN            NaN             NaN  
4          NaN           NaN            NaN             NaN  

[5 rows x 145 columns]

There seem to be no duplicates in Inst --> we can safely merge on column Inst

hospitals DF has 279 rows and 145 columns

In [74]:
# Load hospital geolocations
geodata = pd.read_csv('../data/processed/address_coordinates.csv')
geodata.info()
geodata.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279 entries, 0 to 278
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  279 non-null    int64 
 1   KT          278 non-null    object
 2   Inst        279 non-null    object
 3   Adr         273 non-null    object
 4   Ort         278 non-null    object
 5   Typ         278 non-null    object
 6   Full_adr    279 non-null    object
 7   coords      279 non-null    object
dtypes: int64(1), object(7)
memory usage: 17.6+ KB


Unnamed: 0  KT                         Inst                 Adr  \
0           0  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1           1  AG       Kantonsspital Baden AG            Im Ergel   
2           2  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3           3  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4           4  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ                             Full_adr  \
0        5000 Aarau  K112            Tellstrasse 25 5000 Aarau   
1        5404 Baden  K112                  Im Ergel 5404 Baden   
2        5000 Aarau  K112              Schänisweg 1 5000 Aarau   
3  4310 Rheinfelden  K121  Riburgerstrasse 12 4310 Rheinfelden   
4      5630 Muri AG  K121       Spitalstrasse 144 5630 Muri AG   

                                 coords  
0  ('47.38837665', '8.060297010142904')  
1           ('47.4552657', '8.2820535')  
2           ('47.3911754', '8.0419288')  
3  ('47.55611585', '7.807021507134829')  
4  ('47.27781555', '8.329914342514007')

geodata df has 279 rows (euqal to hospitals DF) and 8 columns. The only column we need for the merge is colum "coords". So we should get a DF with 279 rows and 146 columns after the merge.

In [75]:
# Merge DataFrames
collated_df = hospitals.merge(right=geodata[['Inst', 'coords']], how='left', on='Inst')
collated_df.info()
collated_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 279 entries, 0 to 278
Columns: 146 entries, JAHR to coords
dtypes: float64(133), int64(1), object(12)
memory usage: 318.4+ KB


JAHR  KT                         Inst                 Adr  \
0  2022  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1  2022  AG       Kantonsspital Baden AG            Im Ergel   
2  2022  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3  2022  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4  2022  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ RForm Akt       SL              WB  ...  KostAWLFB  \
0        5000 Aarau  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...        NaN   
1        5404 Baden  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...        NaN   
2        5000 Aarau  K112    R1   A  NF, IPS       Arzt, BGs  ...        NaN   
3  4310 Rheinfelden  K121    R1   A       NF  MSt, Arzt, BGs  ...        NaN   
4      5630 Muri AG  K121    R2   A  NF, IPS  MSt, Arzt, BGs  ...        NaN   

  KostWBLFB KostForLFB  ErlAmbStatLangB  ErlOKPAmbB  ErlKVGStatB  \
0       NaN        NaN              NaN         NaN          NaN   
1       NaN        NaN              NaN         NaN          NaN   
2       NaN        NaN              NaN         NaN          NaN   
3       NaN        NaN              NaN         NaN          NaN   
4       NaN        NaN              NaN         NaN          NaN   

   ErlKVGStatVB  ErlZvOKPStatB  ErlZvOKPStatVB  \
0           NaN            NaN             NaN   
1           NaN            NaN             NaN   
2           NaN            NaN             NaN   
3           NaN            NaN             NaN   
4           NaN            NaN             NaN   

                                 coords  
0  ('47.38837665', '8.060297010142904')  
1           ('47.4552657', '8.2820535')  
2           ('47.3911754', '8.0419288')  
3  ('47.55611585', '7.807021507134829')  
4  ('47.27781555', '8.329914342514007')  

[5 rows x 146 columns]

In [76]:
print(collated_df['coords'].isna().value_counts())

coords
False    279
Name: count, dtype: int64


### Split coordinates into 2 columns called latitude and longitude


In [77]:
split = collated_df['coords'].str.split("'", expand=True)
split.columns = ['bracket', 'latitude', 'coma', 'longitude', 'braket2']
split.head()


bracket     latitude coma          longitude braket2
0       (  47.38837665   ,   8.060297010142904       )
1       (   47.4552657   ,           8.2820535       )
2       (   47.3911754   ,           8.0419288       )
3       (  47.55611585   ,   7.807021507134829       )
4       (  47.27781555   ,   8.329914342514007       )

In [78]:
collated_df[['latitude', 'longitude']] = split[['latitude', 'longitude']]
collated_df.head()

JAHR  KT                         Inst                 Adr  \
0  2022  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1  2022  AG       Kantonsspital Baden AG            Im Ergel   
2  2022  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3  2022  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4  2022  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ RForm Akt       SL              WB  ...  KostForLFB  \
0        5000 Aarau  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   
1        5404 Baden  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   
2        5000 Aarau  K112    R1   A  NF, IPS       Arzt, BGs  ...         NaN   
3  4310 Rheinfelden  K121    R1   A       NF  MSt, Arzt, BGs  ...         NaN   
4      5630 Muri AG  K121    R2   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   

  ErlAmbStatLangB ErlOKPAmbB  ErlKVGStatB  ErlKVGStatVB  ErlZvOKPStatB  \
0             NaN        NaN          NaN           NaN            NaN   
1             NaN        NaN          NaN           NaN            NaN   
2             NaN        NaN          NaN           NaN            NaN   
3             NaN        NaN          NaN           NaN            NaN   
4             NaN        NaN          NaN           NaN            NaN   

   ErlZvOKPStatVB                                coords     latitude  \
0             NaN  ('47.38837665', '8.060297010142904')  47.38837665   
1             NaN           ('47.4552657', '8.2820535')   47.4552657   
2             NaN           ('47.3911754', '8.0419288')   47.3911754   
3             NaN  ('47.55611585', '7.807021507134829')  47.55611585   
4             NaN  ('47.27781555', '8.329914342514007')  47.27781555   

           longitude  
0  8.060297010142904  
1          8.2820535  
2          8.0419288  
3  7.807021507134829  
4  8.329914342514007  

[5 rows x 148 columns]

The collated df has the proper dimensions and the merge was successful. There are no empty values in the coordinates columns.

### Data reduction

Reduce the dataframe to the date we actually need. This requires the following steps:

#### Filter for hospitals providing basic medical care
Reduce hospitals of activity types (column 'Typ') ['K111', 'K112', 'K121', 'K122', 'K123'] only. This is to reduce to hospitals that perform basic medical care only.

In [79]:
# Filter for hospitals providing basic medical care
type = ['K111', 'K112', 'K121', 'K122', 'K123']
collated_basic_care = collated_df[collated_df['Typ'].isin(type)].copy()

collated_basic_care.info()
collated_basic_care.head()

<class 'pandas.core.frame.DataFrame'>
Index: 101 entries, 0 to 250
Columns: 148 entries, JAHR to longitude
dtypes: float64(133), int64(1), object(14)
memory usage: 117.6+ KB


JAHR  KT                         Inst                 Adr  \
0  2022  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1  2022  AG       Kantonsspital Baden AG            Im Ergel   
2  2022  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3  2022  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4  2022  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ RForm Akt       SL              WB  ...  KostForLFB  \
0        5000 Aarau  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   
1        5404 Baden  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   
2        5000 Aarau  K112    R1   A  NF, IPS       Arzt, BGs  ...         NaN   
3  4310 Rheinfelden  K121    R1   A       NF  MSt, Arzt, BGs  ...         NaN   
4      5630 Muri AG  K121    R2   A  NF, IPS  MSt, Arzt, BGs  ...         NaN   

  ErlAmbStatLangB ErlOKPAmbB  ErlKVGStatB  ErlKVGStatVB  ErlZvOKPStatB  \
0             NaN        NaN          NaN           NaN            NaN   
1             NaN        NaN          NaN           NaN            NaN   
2             NaN        NaN          NaN           NaN            NaN   
3             NaN        NaN          NaN           NaN            NaN   
4             NaN        NaN          NaN           NaN            NaN   

   ErlZvOKPStatVB                                coords     latitude  \
0             NaN  ('47.38837665', '8.060297010142904')  47.38837665   
1             NaN           ('47.4552657', '8.2820535')   47.4552657   
2             NaN           ('47.3911754', '8.0419288')   47.3911754   
3             NaN  ('47.55611585', '7.807021507134829')  47.55611585   
4             NaN  ('47.27781555', '8.329914342514007')  47.27781555   

           longitude  
0  8.060297010142904  
1          8.2820535  
2          8.0419288  
3  7.807021507134829  
4  8.329914342514007  

[5 rows x 148 columns]

#### Calculate care days
Calculate care days provided by the hospital based on the formula elaborated in notebook 'hospital_dataset_exploration.ipynb'

In [80]:
collated_basic_care['care_days'] = collated_basic_care['pBettenBelStatA'] * (collated_basic_care['BettenStatA'] * 365) / 100
collated_basic_care.head()

JAHR  KT                         Inst                 Adr  \
0  2022  AG       Kantonsspital Aarau AG      Tellstrasse 25   
1  2022  AG       Kantonsspital Baden AG            Im Ergel   
2  2022  AG      Hirslanden Klinik Aarau        Schänisweg 1   
3  2022  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12   
4  2022  AG         Stiftung Spital Muri   Spitalstrasse 144   

                Ort   Typ RForm Akt       SL              WB  ...  \
0        5000 Aarau  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...   
1        5404 Baden  K112    R1   A  NF, IPS  MSt, Arzt, BGs  ...   
2        5000 Aarau  K112    R1   A  NF, IPS       Arzt, BGs  ...   
3  4310 Rheinfelden  K121    R1   A       NF  MSt, Arzt, BGs  ...   
4      5630 Muri AG  K121    R2   A  NF, IPS  MSt, Arzt, BGs  ...   

   ErlAmbStatLangB ErlOKPAmbB ErlKVGStatB  ErlKVGStatVB  ErlZvOKPStatB  \
0              NaN        NaN         NaN           NaN            NaN   
1              NaN        NaN         NaN           NaN            NaN   
2              NaN        NaN         NaN           NaN            NaN   
3              NaN        NaN         NaN           NaN            NaN   
4              NaN        NaN         NaN           NaN            NaN   

   ErlZvOKPStatVB                                coords     latitude  \
0             NaN  ('47.38837665', '8.060297010142904')  47.38837665   
1             NaN           ('47.4552657', '8.2820535')   47.4552657   
2             NaN           ('47.3911754', '8.0419288')   47.3911754   
3             NaN  ('47.55611585', '7.807021507134829')  47.55611585   
4             NaN  ('47.27781555', '8.329914342514007')  47.27781555   

           longitude  care_days  
0  8.060297010142904   140580.0  
1          8.2820535   116437.0  
2          8.0419288    43434.0  
3  7.807021507134829    37558.0  
4  8.329914342514007    32038.0  

[5 rows x 149 columns]

#### Include required columns only
We only require the following columns: ['KT', 'Inst', 'Adr', 'Ort', 'BettenStatA', 'pBettenBelStatA', 'coords']
In the process rename columns as follows for increased clarity:
pBettenBelStatA --> bed_occupancy
BettenStatA --> max_capacity_in_bed_days

In [81]:
reduced_basic_care = collated_basic_care[['KT', 'Inst', 'Adr', 'Ort', 'latitude', 'longitude', 'care_days', 'BettenStatA', 'pBettenBelStatA',]].copy()

reduced_basic_care.info()
reduced_basic_care.head()

<class 'pandas.core.frame.DataFrame'>
Index: 101 entries, 0 to 250
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   KT               101 non-null    object 
 1   Inst             101 non-null    object 
 2   Adr              100 non-null    object 
 3   Ort              101 non-null    object 
 4   latitude         91 non-null     object 
 5   longitude        91 non-null     object 
 6   care_days        101 non-null    float64
 7   BettenStatA      101 non-null    float64
 8   pBettenBelStatA  101 non-null    float64
dtypes: float64(3), object(6)
memory usage: 7.9+ KB


KT                         Inst                 Adr               Ort  \
0  AG       Kantonsspital Aarau AG      Tellstrasse 25        5000 Aarau   
1  AG       Kantonsspital Baden AG            Im Ergel        5404 Baden   
2  AG      Hirslanden Klinik Aarau        Schänisweg 1        5000 Aarau   
3  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12  4310 Rheinfelden   
4  AG         Stiftung Spital Muri   Spitalstrasse 144      5630 Muri AG   

      latitude          longitude  care_days  BettenStatA  pBettenBelStatA  
0  47.38837665  8.060297010142904   140580.0   493.989041        77.967455  
1   47.4552657          8.2820535   116437.0   364.000000        87.638868  
2   47.3911754          8.0419288    43434.0   155.000000        76.772426  
3  47.55611585  7.807021507134829    37558.0   128.860274        79.852872  
4  47.27781555  8.329914342514007    32038.0   101.000000        86.906280

In [82]:
columns = reduced_basic_care.columns
columns = [word.replace('pBettenBelStatA', 'bed_occupancy') for word in columns]
columns = [word.replace('BettenStatA', 'max_capacity_in_bed_days') for word in columns]
print(columns)

reduced_basic_care.columns = columns
reduced_basic_care.head()

['KT', 'Inst', 'Adr', 'Ort', 'latitude', 'longitude', 'care_days', 'max_capacity_in_bed_days', 'bed_occupancy']


KT                         Inst                 Adr               Ort  \
0  AG       Kantonsspital Aarau AG      Tellstrasse 25        5000 Aarau   
1  AG       Kantonsspital Baden AG            Im Ergel        5404 Baden   
2  AG      Hirslanden Klinik Aarau        Schänisweg 1        5000 Aarau   
3  AG  Gesundheitszentrum Fricktal  Riburgerstrasse 12  4310 Rheinfelden   
4  AG         Stiftung Spital Muri   Spitalstrasse 144      5630 Muri AG   

      latitude          longitude  care_days  max_capacity_in_bed_days  \
0  47.38837665  8.060297010142904   140580.0                493.989041   
1   47.4552657          8.2820535   116437.0                364.000000   
2   47.3911754          8.0419288    43434.0                155.000000   
3  47.55611585  7.807021507134829    37558.0                128.860274   
4  47.27781555  8.329914342514007    32038.0                101.000000   

   bed_occupancy  
0      77.967455  
1      87.638868  
2      76.772426  
3      79.852872  
4      86.906280

### Export dataset

In [83]:
reduced_basic_care.to_csv('../data/processed/hospitals_collated_reduced.csv')